compiler/rustc_middle/src/ty/query.rs

```diff
@@ -231,6 +254,52 @@ macro_rules! define_callbacks {
             })*
         }
 
+        impl<$tcx> TyCtxtFeed<$tcx> {
+            $($(#[$attr])*
+            #[inline(always)]
+            pub fn $name<V>(
```

Comment:
Would it make sense to only allow this for queries which explicitly opt-in? And then maybe not allow setting any provider for such queries. I think that makes the information flow a lot clearer.